### DISCUSSION TITLE
Revisiting automation for closing old pull requests

### DISCUSSION THREAD

**Discussion by errose28** [apache/ozone#7096] [Ideas]
With Ozone's open PR count creeping upwards, I wanted to start a discussion to revisit how we handle old pull requests that have not received updates in a long time. This came up at the US timezone community sync and I wanted to raise it here to a broader audience.

## Motivation

The open pull requests pages and counter is useful for reviewers to gauge whether we are staying on top of incoming changes, something that is always a challenge for a project of this size and complexity. Having too many PRs that remain open but are no longer active, or never leave draft state, adds noise and makes this harder to manage. A look at [page 5](https://github.com/apache/ozone/pulls?page=5&q=is%3Apr+is%3Aopen) of our current PR list shows draft PRs that are pushing two years old without any updates. The bottom of [page 4](https://github.com/apache/ozone/pulls?page=4&q=is%3Apr+is%3Aopen) isn't much better.

## Current System

From our [ci.md](https://github.com/apache/ozone/blob/43244b8bda876f4677abbe576a681436a323e81a/.github/ci.md#close-prs-workflow) documentation:

> close-prs Workflow
[This](https://github.com/apache/ozone/blob/43244b8bda876f4677abbe576a681436a323e81a/.github/workflows/close-pending.yaml) workflow is scheduled each night at midnight; it closes PR's that have not been updated in the last 21 days, while letting the author know they are free to reopen.

However, this description is not quite accurate. For the close workflow to actually close a PR, it needs to have been [manually labelled](https://github.com/apache/ozone/blob/2e329ba3b80d764b24c69292ef4fdcb011b3b4bf/.github/close-pending.sh#L40) as `pending` first. Since most people are not going around manually marking PRs, the workflow does not help us much.

## Proposed Change

Closing a PR is a reversible action to move it out of the review queue. It can be re-opened at any time and no information is lost. In some cases closing works as a ping to reviewers or the contributor to check on tasks that may have been stalled or forgotten. IMO we should not be afraid of closing PRs. Based on this, I think an implementation that meets the following criteria would serve us better:

- Those subscribed to the PR are pinged some amount of time before the PR is closed as a reminder.
  - I propose 21 days before this happens, but the number is somewhat arbitrary.
- The PR is closed some amount of time after the initial reminder if no updates occur.
  - I propose 7 days after the initial reminder, but this number is also arbitrary. This means the PR needs 28 total days of inactivity to be closed.
 - The whole process is automated.  

Luckily GitHub's https://github.com/actions/stale does exactly this with the ability to configure it as needed. **I propose removing our current pending/close workflow and replacing it with actions/stale**.

Let me know your thoughts on this. I wanted to raise this as a discussion instead of a PR so it is seen by a wider audience.

**Comment by adoroszlai** [comment]
I support moving to a standard action.  The rationale for only closing PRs with the `pending` label is to distinguish PRs waiting for action from author vs. from reviewers.

**Reply by errose28** [reply]
Thanks Attila. The intent of the original makes sense in that regard. With actions/stale we will have the option to filter by labelled issues and ping the reviewers if it is blocked on review, which will count as activity and leave the PR open.

**Comment by errose28** [comment]
I filed HDDS-11370 for this. I plan to submit a patch soon.

**Comment by smengcl** [comment]
Closing this thread as [HDDS-11370](https://issues.apache.org/jira/browse/HDDS-11370) is resolved now. Thanks all.
